smart-contracts/contracts/ContentDirectory.sol

```diff
@@ -0,0 +1,475 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.6.0;
+pragma experimental ABIEncoderV2;
+
+// TODO: Interfaces
+import "./storage/ChannelStorage.sol";
+import "./storage/CuratorGroupStorage.sol";
+import "./storage/VideoStorage.sol";
+
+import "./bridge/MembershipBridge.sol";
+import "./bridge/ContentWorkingGroupBridge.sol";
+import "./bridge/auth.sol";
+
+import "../node_modules/@openzeppelin/contracts/utils/Pausable.sol";
+import "../node_modules/@openzeppelin/contracts/access/Ownable.sol";
+
+contract ContentDirectory is RuntimeManageable, Pausable {
+  // Storage contracts
+  ChannelStorage public channelStorage;
+  VideoStorage public videoStorage;
+  CuratorGroupStorage public curatorGroupStorage;
+  // "Bridge" contracts
+  MembershipBridge public membershipBridge;
+  ContentWorkingGroupBridge public contentWorkingGroupBridge;
+
+  // Limits
+  // (they can me "migrated" just be setting a default value here, so no need to keep them in a separate storage)
+  uint64 public channelInstancesLimit = 10;
+  uint32 public videosPerChannelDefaultLimit = 5;
+  uint32 public videosPerChannelMaxLimit = 1000;
+
+  // Enum + constant related to group permissions
+  enum ContentDirectoryOperation {UpdateChannelMetadata, AddVideo, UpdateVideoMetadata, RemoveVideo}
+  uint8 constant GROUP_PERMISSIONS_FLAGS_LENGTH = 4;
+
+  // Reason validation consts
+  uint16 constant DEACTIVATE_CHANNEL_REASON_MIN_LENGTH = 2;
+  uint16 constant DEACTIVATE_VIDEO_REASON_MIN_LENGTH = 2;
+  uint16 constant REMOVE_VIDEO_REASON_MIN_LENGTH = 2;
+
+  // Limits change events
+  event ChannelInstancesLimitUpdated(uint64 _newLimit);
+  event VideosPerChannelDefaultLimitUpdated(uint32 _newLimit);
+  event VideosPerChannelMaxLimitUpdated(uint32 _newLimit);
+
+  // Migration event
+  event Migrated(address _logic, address _videoStorage, address _channelStorage, address _curatorGroupStorage);
+
+  // Channel-related events
+  event ChannelCreated(uint64 _id, ChannelOwnership _ownership, string[2][] _metadata);
```

Comment:
It's a dynamic-size array of arrays containing exactly 2 strings. Initially it was meant to represent cusomizable metadata with dynamic `key-value` pairs, ie.:
```
[
  [ 'title', 'Some video title' ],
  [ 'description', 'Some video description' ]
]
```
But since this is not very flexible, in https://github.com/Joystream/joystream/issues/1816 I explored an alternative possiblity to replace it with just a `string` containing `json` metadata representation (I think that's probably the best way to solve this if the contracts don't need to care about metadata validity and structure)